draft-specifications/Netlayers.md

```diff
@@ -100,6 +100,22 @@ should be hosted on the port `9045` and using "ED25519-V3" for the key type.
 Upon creation of a hidden service tor provides the "Service-ID", this must be
 supplied as the "designator" in the OCapN Node Locator.
 
+# TCP Testing Netlayer
+
+This is a testing-only, simplistic netlayer, providing no security or anonymity
+whatsoever, and only intended to run tests without the huge overhead of setting up
+Tor nodes.
+
+## Implementation
+
+When creating and decoding the OCapN locator, the following information is used:
+
+- **designator**: IPv4 address of a machine. Usually 127.0.0.1 for testing.
+- **transport**: The symbol `tcp`
+- **hints**: Hints are not used, so this may be set to false/omitted.
```

Comment:
I believe this is consistent with the status quo, but I would like to entertain a different design. An OCapN locator needs to capture:

1. One transport-agnostic datum
2. One or more transport-specific data that can be _filtered_ according to their **transport**

I would suggest that the **transport** needs to (at least for now) be much more specific. TCP gives us the transport protocol, but does not specify the message framing protocol or CapTP version. If we want just one `tcp`, we will need a protocol negotiation handshake to allow for the evolution of the protocol, which I think we should avoid for performance reasons. I would prefer to capture all these dimensions in the **transport** datum. In this case, those dimensions are:

1. TCP for **sessions**
2. Syrup for **message framing** (Please confirm.)
3. Syrup **intermediate representation**.
4. OCapN (0) **CapTP protocol**.

By contrast, Agoric supports a number of combinations, but for example:

1. TCP for **session**
2. Netstring for **message frame**
3. JSON for **intermediate representation**
4. Endo/Agoric CapTP (0) for **CapTP**

And also:

1. WebSocket for **session** _and_ **message framing**
2. JSON for **intermediate representation**
3. Endo/Agoric CapTP 0 for **CapTP**

We also support domain sockets and named pipes for **session**, Chrome host messaging for **message framing**, structured clone in some cases instead of JSON, and of course intend to eventually add OCapN to the CapTP layer. We might even also bridge Syrup. We also have the compact-ordered as a candidate for the **intermediate representation** of message bodies. And, the CapTP layer also subsumes the separable concerns of the `{slots, body}` envelope.

To this end, I’m tentatively using exhaustive **transport** terms like `tcp+netstring+json+captp0` and would recommend one like `tcp+syrup+ocapn0` here. I do imagine that as we converge on a small set of good combinations of layers, that we will want to abbreviate to something like `tcp0`, `tcp1`.

To my eye, for TCP, both `tcp+syrup+ocapn0` and the IP address belong in an array (like, **hints**) and the designator should be a transport-agnostic sturdy-ref and whatever cryptography that entails. Tor muddies this with its onion address that _could_ serve as both sturdy-ref and transport-specific address, but I would hope it is transport-specific.

Reply:
well, having proper schemas in protocol part would be incredibly nice, I agree — however I think that someone was against it (I would prefer having `ocapn+onion` and `ocapn+bson+gz+tcp`, instead of weird postfixes we have in the spec rn)

Reply:
cc @dpwiz

Reply:
and bear in mind this is outside of scope of this proposal)

Reply:
and just from the top of my head, it will also present a challenge for negotiating supported protocols. e.g you can connect via `tcp` but `gz` is not implemented.

so yeah, that really does require a full new proposal)